### PR TITLE
Loans: borrow & repay method using pricing-based principal

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

Jeroen's Slack introduction to this feature is [here](https://centrifugedao.slack.com/archives/C05DX0RH1T3/p1688128534804769)

Basically, we want to write the principal in different ways depending on the pricing method. As a schema, instead of always given a `Balance` as principal in `borrow()` and `repay()` methods, we want:
```rust
enum Principal {
    Internal {
        amount: Balance
    },
    External {
        quantity: Rate,
        settlement_price: Balance,
    }
}
```

The core idea is that for external pricing, the amount that is transferred to the pool is `quantity * settlement_price`, and the `outstanding_quantity` is reduced by `quantity`.

So, with this schema:
- Amount transferred: `quantity * settlement_price`
- Amount used for calculating interest: `quantity * notional`
- Amount used for valuating the loan: `quantity * current_price`

## Changes
- [x] Implementation
- [x] Adapt testing
- [x] Adapt benchmarks

### CODE REVIEW

**Comment by lemunozm** [comment]
@mustermeiszer Because we are in a hurry, the main implementation can already be reviewed (it's finished) while I adapt/add tests. 🙌🏻

**Comment by lemunozm** [comment]
Now it's ready for a final review 👐🏻

File: pallets/loans/src/entities/pricing/external.rs [hunk 1]
```diff
@@ -10,14 +10,41 @@ use frame_support::{self, ensure, RuntimeDebug, RuntimeDebugNoBound};
...
 use crate::{
 	entities::interest::ActiveInterestRate,
 	pallet::{Config, Error, PriceOf},
 };
 
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebugNoBound, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct ExternalAmount<T: Config> {
+	pub quantity: T::Rate,
+	pub settlement_price: T::Balance,
```
Comment (mustermeiszer) [inline pallets/loans/src/entities/pricing/external.rs:25]: Nit pick. Not sure if `settlement_price` is the right tearm for when repaying? cc @denniswell

File: pallets/loans/src/entities/pricing/external.rs [hunk 2]
```diff
@@ -127,42 +150,42 @@ impl<T: Config> ExternalActivePricing<T> {
...
-			let price = self.current_price()?;
-
-			let quantity = T::Rate::ensure_from_rational(principal, price)?;
+		self.outstanding_quantity = quantity_adj.ensure_add(self.outstanding_quantity)?;
 
+		let interest_adj = quantity_adj.try_map(|quantity| -> Result<_, DispatchError> {
 			ensure!(
-				quantity.frac().is_zero(),
-				Error::<T>::AmountNotMultipleOfPrice
+				quantity.frac().is_zero() && quantity >= T::Rate::zero(),
```
Comment (mustermeiszer) [inline pallets/loans/src/entities/pricing/external.rs:182]: Again, why the change to allow zero

**Review by mustermeiszer — APPROVED** [review]
I am approving as I spot nothing bad. Just would want to get feedback from Dennis regarding naming and am curious for the relaxation of the `ensure!`-clauses.

**Comment by mustermeiszer** [comment]
Code change of course looks good and clean! :-)

File: pallets/loans/src/entities/pricing/external.rs [hunk 3]
```diff
@@ -46,7 +73,7 @@ impl<T: Config> ExternalPricing<T> {
 	pub fn validate(&self) -> DispatchResult {
 		if let MaxBorrowAmount::Quantity(quantity) = self.max_borrow_amount {
 			ensure!(
-				quantity.frac().is_zero() && quantity > T::Rate::zero(),
+				quantity.frac().is_zero() && quantity >= T::Rate::zero(),
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:76]: It's totally fine to repay with 0 of quantity if you want, for example, to pay only interest.

Or you can create a loan that can not be borrowed only to repay later unchecked amounts.

File: pallets/loans/src/entities/pricing/external.rs (hunk 2 above)
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:182]: It's totally fine to repay with 0 of quantity if you want, for example, to pay only interest.

File: pallets/loans/src/entities/pricing/external.rs (hunk 1 above)
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:25]: As a note: `ExternalAmount` is used for both: borrow & repay. So if changed there, it's changed for both.

Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:25]: From my understanding of what this means, I think the name fits with its purpose.

**Comment by lemunozm** [comment]
Merged, if finally disagree with something. I'll add a PR with that tech debt
